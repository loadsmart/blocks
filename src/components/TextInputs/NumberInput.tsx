import React, { PureComponent } from 'react'
import {
  NativeSyntheticEvent,
  Platform,
  StyleSheet,
  TextInput,
  TextInputKeyPressEventData,
  TextInputProps,
} from 'react-native'
import { Colors } from '../../res'
import { ThemeContext } from '../Contexts'

interface Props extends TextInputProps {
  onChangeText: (text: string) => void
  digit?: string
  focus?: boolean
}

interface State {
  focused: boolean
}

const digitRegex = '^\\d$'

export default class NumberInput extends PureComponent<Props, State> {
  private textInput: TextInput | null

  constructor(props: Props, state: State) {
    super(props, state)
    this.textInput = null
    this.state = {
      focused: false,
    }
  }

  public render() {
    return (
      <ThemeContext.Consumer>
        {theme => {
          const borderStyle = this.state.focused
            ? { borderColor: theme.primaryColor }
            : { borderColor: Colors.Ghost }
          return (
            <TextInput
              ref={component => (this.textInput = component)}
              keyboardType={Platform.OS === 'ios' ? 'number-pad' : 'default'}
              maxLength={1}
              selectionColor={Colors.Transparent}
              style={[styles.container, borderStyle]}
              onChangeText={this.onChangeText}
              onBlur={this.onBlur}
              onFocus={this.onFocus}
              blurOnSubmit={false}
              value={this.props.digit}
              onKeyPress={this.onKeyPress}
              caretHidden={true}
            />
          )
        }}
      </ThemeContext.Consumer>
    )
  }

  public onChangeText = (text: string) => {
    if (text.match(digitRegex)) {
      this.props.onChangeText(text)
    }
  }

  public blur = () => {
    this.textInput!.blur()
  }

  public focus = () => {
    this.textInput!.focus()
  }

  public onBlur = () => {
    this.setState({ focused: false })
  }

  public onFocus = () => {
    this.setState({ focused: true })
  }

  public onKeyPress = (event: NativeSyntheticEvent<TextInputKeyPressEventData>) => {
    const key = event.nativeEvent.key
    if (key.match(digitRegex)) {
      this.props.onChangeText(key)
    } else if (key.match('Backspace')) {
      this.props.onChangeText('')
    }
  }
}

const styles = StyleSheet.create({
  container: {
    width: 45,
    height: 63,
    borderWidth: 2,
    backgroundColor: Colors.White,
    fontSize: 20,
    textAlign: 'center',
    marginHorizontal: 4,
  },
})
