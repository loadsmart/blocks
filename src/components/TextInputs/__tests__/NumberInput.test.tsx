import React from 'react'
import { TextInput as RNTextInput, TextInput } from 'react-native'
import renderer from 'react-test-renderer'
import NumberInput from '../NumberInput'

const mockPlatform = (OS: string, version: number) => {
  jest.resetModules()
  const selectedOS = [OS]
  jest.doMock('Platform', () => ({
    OS,
    select: selectedOS,
    Version: version || undefined,
  }))
}

describe('NumberInput', () => {
  describe('render', () => {
    it('renders with gray border when blurred', () => {
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const input = testRenderer.root.findByType(RNTextInput)
      input.props.onBlur()
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })

    it('renders with green border when focused', () => {
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const input = testRenderer.root.findByType(RNTextInput)
      input.props.onFocus()
      expect(testRenderer.toJSON()).toMatchSnapshot()
    })
  })

  describe('onChangeText', () => {
    it('calls the props onChangeText with valid digit', () => {
      const onChangeTextMock = jest.fn()
      const testRenderer = renderer.create(<NumberInput onChangeText={onChangeTextMock} />)
      testRenderer.root.instance.onChangeText('5')
      expect(onChangeTextMock).toHaveBeenCalled()
    })

    it('does not call the props onChangeText with invalid digit', () => {
      const onChangeTextMock = jest.fn()
      const testRenderer = renderer.create(<NumberInput onChangeText={onChangeTextMock} />)
      testRenderer.root.instance.onChangeText('B')
      expect(onChangeTextMock).not.toHaveBeenCalled()
    })
  })

  describe('onKeyPress', () => {
    it('calls the props onChangeText with valid digit', () => {
      const onChangeTextMock = jest.fn()
      const testRenderer = renderer.create(<NumberInput onChangeText={onChangeTextMock} />)
      const input = testRenderer.root.findByType(RNTextInput)
      input.props.onKeyPress({ nativeEvent: { key: '5' } })
      expect(onChangeTextMock).toHaveBeenCalled()
    })

    it('calls the props onChangeText with empty value', () => {
      const onChangeTextMock = jest.fn()
      const testRenderer = renderer.create(<NumberInput onChangeText={onChangeTextMock} />)
      const input = testRenderer.root.findByType(RNTextInput)
      input.props.onKeyPress({ nativeEvent: { key: 'Backspace' } })
      expect(onChangeTextMock).toHaveBeenCalled()
    })

    it('does not call the props onChangeText with invalid text', () => {
      const onChangeTextMock = jest.fn()
      const testRenderer = renderer.create(<NumberInput onChangeText={onChangeTextMock} />)
      const input = testRenderer.root.findByType(RNTextInput)
      input.props.onKeyPress({ nativeEvent: { key: 'B' } })
      expect(onChangeTextMock).not.toHaveBeenCalled()
    })
  })

  describe('blur', () => {
    it('calls the text input blur function when called', () => {
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const textInput = testRenderer.root.findByType(TextInput)
      textInput.instance.blur = jest.fn()
      testRenderer.root.instance.blur()
      expect(textInput.instance.blur).toHaveBeenCalled()
    })
  })

  describe('focus', () => {
    it('calls the text input focus function when called', () => {
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const textInput = testRenderer.root.findByType(TextInput)
      textInput.instance.focus = jest.fn()
      testRenderer.root.instance.focus()
      expect(textInput.instance.focus).toHaveBeenCalled()
    })
  })

  describe('onBlur', () => {
    it('should update the state to have focused value equals to false', () => {
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const instance = testRenderer.root.instance
      instance.setState({ focused: true })
      instance.onBlur()
      expect(instance.state.focused).toBeFalsy()
    })
  })

  describe('onFocus', () => {
    it('should update the state to have focused value equals to true', () => {
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const instance = testRenderer.root.instance
      instance.setState({ focused: false })
      instance.onFocus()
      expect(instance.state.focused).toBeTruthy()
    })
  })

  describe('keyboardType', () => {
    it('displays default keyboard type on Android', () => {
      mockPlatform('android', 23)
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const input = testRenderer.root.findByType(RNTextInput)
      expect(input.instance.props.keyboardType).toEqual('default')
    })

    it('displays number-pad keyboard type on iOS', () => {
      mockPlatform('ios', 12)
      const testRenderer = renderer.create(<NumberInput onChangeText={jest.fn()} />)
      const input = testRenderer.root.findByType(RNTextInput)
      expect(input.instance.props.keyboardType).toEqual('number-pad')
    })
  })
})
