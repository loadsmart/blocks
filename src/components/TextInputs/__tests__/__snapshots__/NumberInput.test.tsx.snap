// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput render renders with gray border when blurred 1`] = `
<TextInput
  allowFontScaling={true}
  blurOnSubmit={false}
  caretHidden={true}
  keyboardType="number-pad"
  maxLength={1}
  onBlur={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  rejectResponderTermination={true}
  selectionColor="#00000000"
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderWidth": 2,
        "fontSize": 20,
        "height": 63,
        "marginHorizontal": 4,
        "textAlign": "center",
        "width": 45,
      },
      Object {
        "borderColor": "#c4ccd6",
      },
      undefined,
    ]
  }
  textContentType="none"
  underlineColorAndroid="transparent"
/>
`;

exports[`NumberInput render renders with green border when focused 1`] = `
<TextInput
  allowFontScaling={true}
  blurOnSubmit={false}
  caretHidden={true}
  keyboardType="number-pad"
  maxLength={1}
  onBlur={[Function]}
  onChangeText={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  rejectResponderTermination={true}
  selectionColor="#00000000"
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderWidth": 2,
        "fontSize": 20,
        "height": 63,
        "marginHorizontal": 4,
        "textAlign": "center",
        "width": 45,
      },
      Object {
        "borderColor": "#14d64d",
      },
      undefined,
    ]
  }
  textContentType="none"
  underlineColorAndroid="transparent"
/>
`;
